File: src/emberjs/websocket.js

var LibraryEmberWebsocket = {
    $EmberWebsocket: {
        SyncSocket: function(ws) {
            this.ws = ws;
            this.msg_queue = [];
            
            var self = this;

            ws.onopen = function() {
                self.msg_queue.push({type:0});
            };

            ws.onmessage = function(msg) {
                self.msg_queue.push({type:1, message:msg.data});
            };
            
            ws.onclose = function(e) {
                self.msg_queue.push({type:2, message:{code:e.code, reason:e.reason}});
            };
        },
        websockets: {},
        websocket_next: 1,
        ws_open: function(addr, port) {
            var ws = new WebSocket("ws://" + addr + ":" + port + "/");
            this.websockets[this.websocket_next] = new this.SyncSocket(ws);
            return this.websocket_next++;
        },
        ws_poll: function(id) {
            var sock = this.websockets[id];
            var msg = sock.msg_queue.shift();
            if (msg === undefined) {
                return null;
            } else {
                return msg;
            }
        },
        ws_send: function(id, msg) {
            var sock = this.websockets[id];
            sock.ws.send(msg);
        },
        ws_close: function(id) {
            var sock = this.websockets[id];
            sock.ws.close();
            delete this.websockets[id];
        }
    },
    
    ember_ws_open: function(addr, port) {
        return EmberWebsocket.ws_open(Module.Pointer_stringify(addr), port);
    },
    ember_ws_poll: function(id) {
        var msg = EmberWebsocket.ws_poll(id);
        if (msg === null) {
            return 0;
        } else {
            var str = JSON.stringify(msg);
            var len = lengthBytesUTF8(str)+1;
            var mem = _malloc(len);
            stringToUTF8(str, mem, len);
            return mem;
        }
    },
    ember_ws_send: function(id, msg) {
        EmberWebsocket.ws_send(id, Module.Pointer_stringify(msg));
    },
    ember_ws_close: function(id) {
        EmberWebsocket.ws_close(id);
    },
};

autoAddDeps(LibraryEmberWebsocket, '$EmberWebsocket');
mergeInto(LibraryManager.library, LibraryEmberWebsocket);

File: src/emberjs/websocket.cpp
#include "websocket.hpp"

#include "../utility.hpp"

using json = nlohmann::json;

namespace emberjs {

    void from_json(const json& j, event_type& type) {
        type = static_cast<event_type>(j.get<int>());
    }

    websocket::websocket(const std::string& address, int port) :
        id(ember_ws_open(address.c_str(), port))
    {}

    websocket::websocket(websocket&& other) :
        id(std::exchange(other.id, 0)),
        message_callback(std::move(other.message_callback))
    {}

    websocket& websocket::operator=(websocket&& other) {
        this->~websocket();
        new (this) websocket(std::move(other));
        return *this;
    }

    websocket::~websocket() {
        ember_ws_close(id);
    }

    void websocket::on_open(std::function<void()> callback) {
        open_callback = std::move(callback);
    }

    void websocket::on_message(std::function<void(const std::string&)> callback) {
        message_callback = std::move(callback);
    }

    void websocket::on_close(std::function<void(const std::string&)> callback) {
        close_callback = std::move(callback);
    }

    void websocket::poll() {
        while (auto msg = ember_ws_poll(id)) {
            EMBER_DEFER { free(msg); };
            auto jmsg = json::parse(msg);
            switch ((event_type)jmsg["type"]) {
                case event_type::OPEN:
                    open_callback();
                    break;
                case event_type::MESSAGE:
                    message_callback(jmsg["message"]);
                    break;
                case event_type::CLOSE:
                    close_callback(jmsg["message"]);
                    break;
            }
        }
    }

    void websocket::send(const std::string msg) {
        ember_ws_send(id, msg.c_str());
    }

} //namespace emberjs
